group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetricsShard.java

```diff
@@ -0,0 +1,325 @@
+        if (sensor != null) {
+            sensor.record(val);
+        }
+    }
+
+    @Override
+    public TopicPartition topicPartition() {
+        return this.topicPartition;
+    }
+
+    @Override
+    public void commitUpTo(long offset) {
+        this.localGauges.forEach((__, gaugeCounter) -> {
+            long value;
+            synchronized (gaugeCounter.timelineLong) {
+                value = gaugeCounter.timelineLong.get(offset);
+            }
+            gaugeCounter.atomicLong.set(value);
+        });
+    }
+
+    /**
+     * Called when a generic group's state has changed. Increment/decrement
+     * the counter accordingly.
+     *
+     * @param oldState The previous state. null value means that it's a new group.
+     * @param newState The next state. null value means that the group has been removed.
+     */
+    public void onGenericGroupStateTransition(GenericGroupState oldState, GenericGroupState newState) {
+        if (newState != null) {
+            switch (newState) {
+                case PREPARING_REBALANCE:
+                    incrementGlobalGauge(NUM_GENERIC_GROUPS_PREPARING_REBALANCE);
+                    break;
+                case COMPLETING_REBALANCE:
+                    incrementGlobalGauge(NUM_GENERIC_GROUPS_COMPLETING_REBALANCE);
+                    break;
+                case STABLE:
+                    incrementGlobalGauge(NUM_GENERIC_GROUPS_STABLE);
+                    break;
+                case DEAD:
+                    incrementGlobalGauge(NUM_GENERIC_GROUPS_DEAD);
+                    break;
+                case EMPTY:
+                    incrementGlobalGauge(NUM_GENERIC_GROUPS_EMPTY);
+            }
+        } else {
+            decrementLocalGauge(NUM_GENERIC_GROUPS);
+        }
+
+        if (oldState != null) {
+            switch (oldState) {
+                case PREPARING_REBALANCE:
+                    decrementGlobalGauge(NUM_GENERIC_GROUPS_PREPARING_REBALANCE);
+                    break;
+                case COMPLETING_REBALANCE:
+                    decrementGlobalGauge(NUM_GENERIC_GROUPS_COMPLETING_REBALANCE);
+                    break;
+                case STABLE:
+                    decrementGlobalGauge(NUM_GENERIC_GROUPS_STABLE);
+                    break;
+                case DEAD:
+                    decrementGlobalGauge(NUM_GENERIC_GROUPS_DEAD);
+                    break;
+                case EMPTY:
+                    decrementGlobalGauge(NUM_GENERIC_GROUPS_EMPTY);
+            }
+        } else {
+            incrementLocalGauge(NUM_GENERIC_GROUPS);
+        }
+    }
+
+    /**
+     * Called when a consumer group's state has changed. Increment/decrement
+     * the counter accordingly.
+     *
+     * @param oldState The previous state. null value means that it's a new group.
+     * @param newState The next state. null value means that the group has been removed.
+     */
+    public void onConsumerGroupStateTransition(
```

Comment:
hmm -- given the states will be different -- should we block this change on the state changes in https://github.com/apache/kafka/pull/14673?

Reply:
https://github.com/apache/kafka/pull/14673 changes the MemberState whereas this PR we only record metrics for ConsumerGroupStates

Reply:
whoops. ok cool so no blocking needed.